Get ObjectId from timestamp

rvazquezglez opened the pull request:
As mentioned [here](https://steveridout.github.io/).

# Why generate an ObjectId from a timestamp?

To query documents by creation date.

e.g. to find all comments created after 2013-11-01:

 ```
 db.comments.find({_id: {$gt: ObjectId("5272e0f00000000000000000")}})
 ```
 The intention is to use the new function in SQL like:
  ```
  SELECT * 
  FROM comments 
  WHERE '_id' > timestamp_objectid(TIMESTAMP '2013-11-01 00:00:00 +00:00')
  ```
  
  (`_id` being the ObjectId of the document collection)

ebyhr commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestObjectIdFunctions.java:
```diff
@@ -42,6 +45,24 @@ public void testObjectidTimestamp()
+    public void testTimestampObjectid()
+    {
+        assertFunction(
+                "timestamp_objectid((TIMESTAMP '1979-09-05 22:51:36 +00:00'))",
```
Please remove a redundant parentheses and add another case with `null`.

rvazquezglez replied:
Removed redundant parentheses and added a test case with `null`.

ebyhr commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/ObjectIdFunctions.java:
```diff
@@ -60,6 +63,15 @@ public static long objectidTimestamp(@SqlType("ObjectId") Slice value)
+    @Description("Mongodb ObjectId from the given timestamp")
+    @ScalarFunction
+    @SqlType("ObjectId")
+    public static Slice timestampObjectid(@SqlType("timestamp(3) with time zone") long value)
```
Rename `value` to `timestamp`.

electrum commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/ObjectIdFunctions.java:
```diff
@@ -60,6 +63,15 @@ public static long objectidTimestamp(@SqlType("ObjectId") Slice value)
+    @Description("Mongodb ObjectId from the given timestamp")
+    @ScalarFunction
+    @SqlType("ObjectId")
+    public static Slice timestampObjectid(@SqlType("timestamp(3) with time zone") long value)
```
This should probably be timestamp(0) since the function doesn’t use the fractional seconds.

rvazquezglez replied:
Thank you! I was not aware of the meaning, I just took the type from the other functions in the class.
Updated it to be timestamp(0)

ebyhr commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/ObjectIdFunctions.java:
```diff
@@ -60,6 +63,15 @@ public static long objectidTimestamp(@SqlType("ObjectId") Slice value)
+    @Description("Mongodb ObjectId from the given timestamp")
+    @ScalarFunction
+    @SqlType("ObjectId")
+    public static Slice timestampObjectid(@SqlType("timestamp(0) with time zone") long value)
```
Rename `value` to `timestamp`.

rvazquezglez replied:
Renamed.

ebyhr reviewed (commented):
Could you update below docs? 
- `docs/src/main/sphinx/connector/mongodb.rst`
- `docs/src/main/sphinx/functions/list.rst`

hashhar commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/ObjectIdFunctions.java:
```diff
@@ -52,6 +54,7 @@ public static Slice objectid(@SqlType(StandardTypes.VARCHAR) Slice value)
 
+    @Description("Timestamp from the given Mongodb ObjectId")
     @ScalarFunction
     @SqlType("timestamp(3) with time zone") // ObjectId's timestamp is a point in time
```
Should the return type be `timestamp(0) with time zone` here too?

rvazquezglez replied:
Makes sense, I've changed it.

ebyhr replied:
@rvazquezglez Please revert the change. Changing return type might be breaking change. We don't want to handle in this PR.

rvazquezglez replied:
Reverted.

rvazquezglez commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestObjectIdFunctions.java:
```diff
@@ -42,6 +45,24 @@ public void testObjectidTimestamp()
+                "ObjectId('12 34 56 78 90 ab cd ef   12 34 56 78')",
+                OBJECT_ID,
+                new SqlVarbinary(new ObjectId("1234567890abcdef12345678").toByteArray()));
+    }
```
I added the test to understand how `objectid(..)` is working and be able to add the new function. I've moved the test to the beginning.

rvazquezglez replied:
Added a separate Pull Request: https://github.com/trinodb/trino/pull/8828

rvazquezglez commented:
@ebyhr I've updated 

- `docs/src/main/sphinx/connector/mongodb.rst`
- `docs/src/main/sphinx/functions/list.rst`

To include information about the timestamp functions and, as an extra, added info about mapping nested fields in the schema.

rvazquezglez commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestObjectIdFunctions.java:
```diff
@@ -42,6 +45,24 @@ public void testObjectidTimestamp()
+    public void testObjectid()
+    {
+        assertFunction(
+                "ObjectId('12 34 56 78 90 ab cd ef   12 34 56 78')",
```
The spaces are part of the test, to check that ObjectId is capable of handling them, I've added a separate test without them.

ebyhr commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestObjectIdFunctions.java:
```diff
@@ -33,17 +36,53 @@ protected void registerFunctions()
     private SqlTimestampWithTimeZone toTimestampWithTimeZone(ZonedDateTime zonedDateTime)
     {
-        return SqlTimestampWithTimeZone.newInstance(3, zonedDateTime.toInstant().toEpochMilli(), 0, TimeZoneKey.getTimeZoneKey(zonedDateTime.getZone().getId()));
+        return SqlTimestampWithTimeZone.newInstance(0, zonedDateTime.toInstant().toEpochMilli(), 0, TimeZoneKey.getTimeZoneKey(zonedDateTime.getZone().getId()));
```
Revert.

rvazquezglez replied:
Reverted.

ebyhr commented on docs/src/main/sphinx/connector/mongodb.rst:
```diff
@@ -244,6 +244,64 @@ Field           Required  Type      Description
+      }
+    }
+
+You can define the nesting by using `row` as follows:
```
Fix invalid format. This is CI failure cause. 
```suggestion
You can define the nesting by using ``row`` as follows:
```

ebyhr commented on docs/src/main/sphinx/connector/mongodb.rst:
```diff
@@ -294,6 +352,41 @@ You can render the ``_id`` field to readable values with a cast to ``VARCHAR``:
+
+    SELECT *
+    FROM collection
+    WHERE '_id' > timestamp_objectid(TIMESTAMP '2021-08-07 17:51:36 +00:00');
```
Single quotation expects a literal, not a column. 
```suggestion
    WHERE  _id > timestamp_objectid(TIMESTAMP '2021-08-07 17:51:36 +00:00');
```

rvazquezglez replied:
removed quotes

ebyhr commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestObjectIdFunctions.java:
```diff
@@ -33,17 +36,53 @@ protected void registerFunctions()
         assertFunction(
                 "objectid_timestamp(ObjectId('1234567890abcdef12345678'))",
-                TIMESTAMP_TZ_MILLIS,
+                TIMESTAMP_TZ_SECONDS,
```
Revert.

rvazquezglez replied:
Reverted

rvazquezglez commented on docs/src/main/sphinx/connector/mongodb.rst:
```diff
@@ -244,6 +244,64 @@ Field           Required  Type      Description
+
+
+Nested fields
+^^^^^^^^^^^^^
```
Added a separate PR https://github.com/trinodb/trino/pull/8829

rvazquezglez commented on docs/src/main/sphinx/connector/mongodb.rst:
```diff
@@ -294,6 +352,41 @@ You can render the ``_id`` field to readable values with a cast to ``VARCHAR``:
+.. code-block:: sql
+
+    SELECT '_id' as ID, objectid_timestamp('_id') as TS
+    FROM collection;
```
Changed to literal example.

ebyhr reviewed (approved):
Looks good except for doc. Please squash commits into one and fix the commit subject & body:
```
Add timestamp_objectid function in MongoDB connector

Additionally, add description and documentation for 
objectid_timestamp function.
```

rvazquezglez commented on docs/src/main/sphinx/connector/mongodb.rst:
```diff
@@ -294,6 +294,40 @@ You can render the ``_id`` field to readable values with a cast to ``VARCHAR``:
+^^^^^^^^^^^^^^^^^^^^^^
+
+Creates an ObjectId from a given timestamp. In MongoDB you can filter all the documents created after August 7th, 2021
+with a query like this:
```
Nice, looks better, I applied them.

ebyhr commented:
Merged, thanks!
